### ISSUE TITLE
[Feature] Expose something like custom VJP in Python

### ISSUE THREAD

**Issue by awni** [ml-explore/mlx#1090]
Consider allowing a function to have a custom VJP function attached to it.

**Comment by angeloskath** [comment]
Oops I thought I had exposed that. It might be worth having a high level API for that as well (which is probably why I didn't expose it), maybe implemented in python using the `custom_vjp` transform? Either of the following two feels more natural to use imho

Option a:
```python
import mlx.core as mx

@mx.custom_function
def my_addition(a, b):
    return a + b

@my_addition.vjp
def my_addition_vjp(primals, outputs, cotangents):
    a, b = primals
    cotan = cotangents[0]
    return cotan * b, cotan * a
```

Option b:
```python
import mlx.core as mx

@mx.custom_function
def my_addition(a, b):
    def my_vjp(primals, outputs, cotangents):
        a, b = primals
        cotan = cotangents[0]
        return cotan * b, cotan * a
    return a + b, my_vjp
```

I lean towards the first one. Obviously names and details need some changing.

**Comment by kaeru-shigure** [comment]
I think the interface your propose is good, but if it becomes pytorch compatible, it might be useful when migrating.  
Anyway, I am eagerly waiting for this feature to be implemented.  
  
Here is an example of a [torch.autograd.Function](https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function) compatible interface:
```py
import mlx.core as mx

class MxCustomVjp():
    @classmethod
    def apply(cls, *args, **kwargs):
        ctx = cls()
        ctx.save_for_backward(*args)
        r = ctx.forward(ctx, *args, **kwargs)
        return r if type(r) == tuple else (r,), ctx
    def save_for_backward(self, *args):
        self.saved_tensors = args
    def vjp(self, grad_output):
        return self.__class__.backward(self, *grad_output)
    @staticmethod
    def forward(self, *args, **kwargs):
        raise NotImplementedError
    @staticmethod
    def backward(self, *args, **kwargs):
        raise NotImplementedError

class CustomLinear(MxCustomVjp):
    @staticmethod
    def forward(ctx, x, weight):
        ctx.save_for_backward(x, weight)
        return x @ weight
    @staticmethod
    def backward(ctx, grad_output):
        x, weight = ctx.saved_tensors
        grad_x = grad_output @ weight.T
        grad_weight = x.T @ grad_output
        return grad_x, grad_weight

pair2x = lambda a, b: (a * 2, b * 2,)
mxMean = lambda a: a.mean()
def test_forward(x, weight):
    r_pair2x = pair2x(x, weight)
    r_lin, ctx_lin = CustomLinear.apply(*r_pair2x)
    r_mean = mxMean(*r_lin)
    # r_pair2x is not used in backward
    backward_bucket = (r_lin, ctx_lin)
    return r_mean, backward_bucket

# I want same process with autograd
def manual_value_and_grad(x, weight):
    # forward
    r, backward_bucket = test_forward(x, weight)
    # backward
    _, grad_mean = mx.vjp(mxMean, backward_bucket[0], (mx.ones_like(r),))
    grad_lin = backward_bucket[1].vjp(grad_mean)
    _, grad = mx.vjp(pair2x, (x, weight,), grad_lin)
    return r, grad

x = mx.random.normal((1,3))
weight = mx.random.normal((x.shape[-1],3))
(r, _), grad = mx.value_and_grad(test_forward, [0, 1])(x, weight)
r_m, grad_m = manual_value_and_grad(x, weight)
assert mx.allclose(r, r_m).item()
assert mx.allclose(grad[0], grad_m[0]).item()
assert mx.allclose(grad[1], grad_m[1]).item()
```

**Comment by awni** [comment]
@kemchenj we have a [custom_vjp transformation](https://github.com/ml-explore/mlx/blob/main/mlx/transforms.h#L181-L190) in C++.

**Comment by awni** [comment]
@angeloskath regarding the Python interface. I think in option A is nicer for enabling any transformation to be customized which I think is preferable.

I'm not so crazy about the name `custom_function`. Maybe `customizable` or even `extension`.  Does this make sense:

```python
my_extendable_fun = mx.extension(my_fun)
my_extendable_fun.vjp = ...
my_extendable_fun.eval_cpu = ...
my_extendable_fun.vmap = ...
```

Presumably it will fall back to the default VJP / other transforms if they are not implemented?

**Comment by awni** [comment]
This is available in v0.16. See the [documentation for more information](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.custom_function.html). 

If you have any feedback or run into any issues please let us know as this is a new feature there may be some kinks that need to be ironed out.
